### ISSUE TITLE
Battery usage

### ISSUE THREAD

**Issue by Rudloff** [streetcomplete/StreetComplete#2439]
I opened a similar issue a while ago (https://github.com/streetcomplete/StreetComplete/issues/1036), but the app still uses a lot of battery on my Fairphone 2 when I am activaly using it. 
The FP2 has a terrible battery life but it is far worse with StreetComplete than with other apps. (Like I can see the battery percentage dropping every minute or so while I am mapping.)

Is there any useful info I can provide to help diagnose this?

I had a look at the CPU usage and it is around 10-30% when the app is in the foreground doing nothing and can go up to 130% when actively interacting with it.

**How to Reproduce**
Have the app in the foreground, pan the map and solve quests.

**Versions affected**
I can reproduce on 28.0 but this is not new behavior.
I am using Android 7.1.2.

**Comment by westnordost** [comment]
Hm that's too much. It should be around 0% if you do nothing. If you do nothing, what's the CPU usage?

**Comment by Rudloff** [comment]
I tried leaving the app open and doing absolutely nothing.
CPU usage reported by `top` stayed between 10 and 20%:
```
  PID USER     PR  NI VIRT  RES  SHR S[%CPU] %MEM     TIME+ ARGS                                                                                                                                                    
 2160 u0_a100  10 -10 1.1G 172M  74M S 16.4   9.1   0:12.74 de.westnordost.streetcomplete                                               
```

Then when my screen went off, it went down to 0:
```
  PID USER     PR  NI VIRT  RES  SHR S[%CPU] %MEM     TIME+ ARGS                                                                                                                                                    
 2160 u0_a100  10 -10 1.0G 173M  77M S  0.0   9.2   0:51.87 de.westnordost.streetcomplete                                               
```

**Comment by westnordost** [comment]
PRetty sure that's tangram-es. Let's see

**Comment by westnordost** [comment]
Hmm,  I can't reproduce this, for me the CPU is on around 1-2% for the app. Maybe you can help here. You could install Android Studio and within Android Studio, there is a builtin profiler. There, you are able to sample what causes the issue.

**Comment by westnordost** [comment]
![image](https://user-images.githubusercontent.com/4661658/103419589-fd7c7580-4b93-11eb-98dc-a252d04b2a18.png)

![image](https://user-images.githubusercontent.com/4661658/103419655-2d2b7d80-4b94-11eb-978e-50600d3dac4b.png)

![image](https://user-images.githubusercontent.com/4661658/103419677-42081100-4b94-11eb-98fd-f988f0f934fa.png)

![image](https://user-images.githubusercontent.com/4661658/103419697-6532c080-4b94-11eb-924a-36f52f17c22a.png)

And then you can record/sample methods 
![image](https://user-images.githubusercontent.com/4661658/103419724-84315280-4b94-11eb-8a17-db9c0a6abb52.png)

Record for a few seconds, then when you stop sampling, there you'll see the results in various different illustrations. The most informative is in my opinion the flame chart. However, you can also save the recording and send it to me:

![image](https://user-images.githubusercontent.com/4661658/103419825-f73ac900-4b94-11eb-97e9-03713c1f7f68.png)

**Comment by westnordost** [comment]
Hmm, and to be able to debug, you need to run a debug version of the app. If you have Android Studio installed anywhere, you could compile it yourself (just load the project, select your phone as deploy target and press on the green bug icon at the top) or if there are any problems, I could build one for you.

**Comment by Rudloff** [comment]
Here is a recording: [cpu-art-20201231T183031.trace.zip](https://github.com/streetcomplete/StreetComplete/files/5757896/cpu-art-20201231T183031.trace.zip)

I am not a Java dev so I am not sure how to read the result.

> Hmm, and to be able to run the profiler, you need to run a debug version of the app.

It seems to run correctly with the production APK from F-droid (but maybe it gives less information?).

**Comment by westnordost** [comment]
Hmm, I looked at the recording and there is nothing conspicuous. I can't believe that this is supposed to be 10% CPU

**Comment by westnordost** [comment]
I also tested it with my older device, a Sony Xperia Z1 Compact, Android 5.1.

If the app is in foreground and the map is not moved, the CPU is busy by 0-10% with the app. If I scroll around, it's about 20%. Doesn't feel that much.

So on idling on the map, what is happening is that the compass sensor is queried to see and render on the map where the user is looking. So the 0-10% is both the compass sensor thread and the OpenGL thread used by the map renderer to paint the view direction.

All in all, I don't feel there is a performance problem. Maybe there are other processes causing the power drain. For the Sony Xperia Z1 Compact, this is how the CPU graph looks while the app is idling in the foreground:

![image](https://user-images.githubusercontent.com/4661658/103485435-927ea900-4df6-11eb-97e9-3ab00bd24557.png)

green is the CPU used by the app, dark bluish is the CPU used by other processes on the device that are not currently in foreground (no idea what is causing this, the phone is basically factory-reset).

**Comment by westnordost** [comment]
And this is how it looks like for my new phone (Samsung S10e):
![image](https://user-images.githubusercontent.com/4661658/103485540-5435b980-4df7-11eb-9906-3c8e2ce66091.png)

Though note that the S10e has
- like the Fairphone 3, an Octa-Core processor. The Sony Xperia has a Quad-Core. 100% CPU in the graph would mean AFAIK that all cores are on 100%
- The Samsung has a 7nm architecture, the Sony has a 24nm architecture

**Comment by HolgerJeromin** [comment]
> like the Fairphone 3

He has a Fairphone 2.

**Comment by Helium314** [comment]
I did a short test on my S4 mini (using an overlay that shows power consumption and per-core frequency)
After inital startup is finished, StreetComplete uses about as much power as simply showing the home screen.
But when no location is available (disabled or no GPS signal), one core runs at highest frequency until StreetComplete has a location. So if you have really bad GPS or location switched off, this could be a problem.

For me disabling auto-upload helped a lot, not sure if it was because of the much less active data connection or because there is less work on the database done in the background.

**Comment by westnordost** [comment]
So if location is disabled, the power usage goes up? That sounds fishy.

Cannot reproduce this though. If location is off, the app dues use even a little less CPU (0-1% rather than 2-3%)

**Comment by Helium314** [comment]
> So if location is disabled, the power usage goes up?

Only if location is disabled when SC is started. If I disable location after SC got at least one location, there is no problem.
Can reproduce this on another device (S4 mini VE, different CPU). Here one core runs at 1 GHz until I get a location, after that it switches to 200 MHz. The core can go up to 1.4 GHz, so it's not fully loaded. I assume on most newer phones you would not even notice this...

Anyway, I doubt that anyone would voluntarily use SC with location disabled so this observation is probably irrelevant.

**Comment by westnordost** [comment]
> Only if location is disabled when SC is started.

Hm, also can't reproduce this.

**Comment by westnordost** [comment]
Incidentally, while looking for another issue, I could suddenly reproduce this myself once. CPU was at 30% all the time and recording CPU usage didn't show anything special other than the usual (compass thread and GL update).

In a later attempt, I couldn't reproduce it anymore and CPU was back at 5%.

But anyway, I optimized a few things about the compass. The sensor data -> rotation calculation is now done half as many times, plus some other minor things. However, it didn't change anything on the 5% CPU though.

I also turned off the compass completely, this made the CPU go to about 4%. Hm ok. 🤷
